### PR TITLE
[LIVE-9653][LIVE-9691] Bugfix - Fix incorrect token id encoding in account ids & incorrect BSC tokens type

### PR DESCRIPTION
**Pull request by lambertkevin**
### 📝 Description

This PR is fixing 2 problems:
-  the token ids encoding in `TokenAccounts` ids could contain characters not handled by the `
encodeURIComponent` (like `_` `-` or `+`), which would then break its "deserialization". 
    - https://github.com/LedgerHQ/ledger-live/blob/develop/libs/coin-framework/src/account/accountId.ts#L36 
    - https://github.com/LedgerHQ/ledger-live/blob/develop/libs/coin-framework/src/operation.ts#L70
    -  https://github.com/LedgerHQ/ledger-live/blob/develop/libs/coin-framework/src/operation.ts#L93
- the BSC tokens were added to the CAL using the `convertERC20` method instead of the `convertBEP20` method, changing the ids from `bsc/bep20/XYZ` to `bsc/erc20/XYZ` which was breaking the mappings we have in some Ledger backends. ⚠️ Having 2 disctinct types for `erc20` and `bep20` is technically **useless**, as they're the same thing. This should be removed ASAP when the backends are capable of doing the migration of ids on their side.
 
### ❓ Context

- **Impacted projects**: `@ledgerhq/coin-framework` `@ledgerhq/coin-evm` 
- **Linked resource(s)**: 
- https://ledgerhq.atlassian.net/browse/LIVE-9653
- https://ledgerhq.atlassian.net/browse/LIVE-9691

### ✅ Checklist

- [x] **Test coverage** 
- [x] **Atomic delivery** 
- [x] **No breaking changes** 

### 🚀 Expectations to reach

_Please make sure you follow these [**Important Steps**](https://github.com/LedgerHQ/ledger-live/blob/develop/CONTRIBUTING.md#important-steps)._

_Pull Requests must pass the CI and be internally validated in order to be merged._

### CODE REVIEW

File: libs/coin-framework/src/account/accountId.ts [hunk 1]
```diff
@@ -11,6 +11,43 @@ function ensureNoColon(value: string, ctx: string): string {
...
+  const URIEncoded = encodeURIComponent(tokenId);
+  const dashUnderscoreSafe = URIEncoded.replace(/-/g, "~!dash!~")
+    .replace(/_/g, "~!underscore!~")
+    .replace(/\+/g, "~!plus!~");
+
+  invariant(
+    !dashUnderscoreSafe.includes("-") &&
+      !dashUnderscoreSafe.includes("_") &&
+      !dashUnderscoreSafe.includes("+"),
+    `TokenId ${dashUnderscoreSafe} shouldn't contain - or _`,
```
Comment (gre) [inline libs/coin-framework/src/account/accountId.ts:26]: you can remove this invariant. the code before it makes this case impossible. that way we can optim a bit the perf (includes can be costly as i assume we call this function a lot)

File: libs/coin-framework/src/account/accountId.ts [hunk 2]
```diff
@@ -11,6 +11,43 @@ function ensureNoColon(value: string, ctx: string): string {
...
+    .replace(/~!underscore!~/g, "_")
+    .replace(/~!plus!~/g, "+");
+  const decodedURIComponent = decodeURIComponent(dashUnderscoreUnsafe);
+
+  invariant(
+    !decodedURIComponent.includes("~!dash!~") &&
+      !decodedURIComponent.includes("~!underscore!~") &&
+      !decodedURIComponent.includes("~!plus!~"),
+    `TokenId ${decodedURIComponent} shouldn't contain ~!dash!~, ~!underscore!~ or ~!plus!~`,
+  );
```
Comment (gre) [inline libs/coin-framework/src/account/accountId.ts:46]: same

**Review by gre — COMMENTED** [review]
ugly but necessary fix. I don't see another solution at the moment. 😢 

I think we shouldn't have the invariant in the code runtime because it doesn't help perf & also it's provable by the code that these are never possible (whatever the user's input param are), you have test that also will help on coverage

File: libs/coin-framework/src/account/accountId.ts (hunk 1 above)
Comment (lambertkevin) [reply libs/coin-framework/src/account/accountId.ts]: Alright, done 👌

File: libs/coin-framework/src/account/accountId.ts (hunk 2 above)
Comment (lambertkevin) [reply libs/coin-framework/src/account/accountId.ts]: Also done 👌

**Review by haammar-ledger — COMMENTED** [review]
@lambertkevin Maybe `encodeCurrencyAsLedgerId` to cover too in `libs/ledger-live-common/src/countervalues/helpers.ts`?

**Comment by lambertkevin** [comment]
> encodeCurrencyAsLedgerId

No idea what this thing is, I'll have a look 👌

**Comment by lambertkevin** [comment]
> LGTM but regarding this:
> 
> > Ok just had a look, and this encodePairAsLedgerIdPair should not be impacted by the issue I'm fixing I guess because It's never "decoded" nor included in something that would then be splitted like the operation Ids or account Ids
> 
> Don't we want to be future-proof? Since it's the same situation that lead us here of the problem being invisible as long as it was never decoded. That said I didn't investigate at all what it's used for, and probably not an issue today indeed according to what you say.

@haammar-ledger Well the problem here was only happening because of this double serialization: a token id was "encoded", then added to a serialized `TokenAccount` id, which was also serialized into an `Operation` id. Going full reverse on this was problematic because you were deserializing something with a `string.split("X")`, and if that `X` was not safely removed during the encoding, then your whole deserialization was fucked, and you were splitting the string in the wrong places. 
As long as you don't have this "split" mecanism, using -_+ or anything in the token id is fine I guess.

So in the case of `encodePairAsLedgerIdPair`, it looks like it's a function used for counter values, and this time it's litteraly added to an URI:

```typescript
const latest = async (pairs: TrackingPair[], direct?: boolean) => {
  const all = await promiseAllBatched(4, chunk(pairs, LATEST_CHUNK), async partial => {
    const url = `${baseURL()}/v2/latest/${direct ? "direct" : "indirect"}?pairs=${partial
      .map(p => encodePairAsLedgerIdPair(p.from, p.to))
      .join(",")}`;
    const { data } = await network({ method: "GET", url });
    return data;
  });
  const data = all.reduce((acc, data) => acc.concat(data), []);
  return data;
};
```

So my guess is that changing anything in the encoding of `encodePairAsLedgerIdPair` would actually make the "pairs" break on the backend side, as they wouldn't know they need to replace the `~!dash!~` with a `-` etc.

**Pull request merged**